src/transports/batch.rs

```diff
@@ -0,0 +1,98 @@
+//! Batching Transport
+
+use std::mem;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use futures::{self, future, Future, BoxFuture};
+use futures::sync::oneshot;
+use parking_lot::Mutex;
+use rpc;
+use transports::Result;
+use {BatchTransport, Transport, Error as RpcError, RequestId};
+
+type Pending = oneshot::Sender<Result<rpc::Value>>;
+
+/// Transport allowing to batch queries together.
+pub struct Batch<T> {
+  transport: T,
+  pending: Arc<Mutex<BTreeMap<RequestId, Pending>>>,
+  batch: Mutex<Vec<(RequestId, rpc::Call)>>,
```

Comment:
i guess it's a slight optimization that you don't have to reallocate the pending set, but it also means that queuing a new batch will be blocked while processing the response of the last one.